resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala

```diff
@@ -158,6 +158,7 @@ private[yarn] class YarnAllocator(
   private[yarn] val containerPlacementStrategy =
     new LocalityPreferredContainerPlacementStrategy(sparkConf, conf, resource, resolver)
 
+  private[yarn] val completedContainerIdSet: HashSet[ContainerId] = new HashSet[ContainerId]()
```

Comment:
Any recycling strategy？I wonder there may be subtle issue of memory consuming.

Reply:
My friend told me the completedContainer will only be reported twice, if that's true, we could use `completedContainerIdSet.remove(containerId)` instead of `completedContainerIdSet.contains(containerId)` as I said above.

But I can't confirm that, I don't know much about Yarn.